### PR TITLE
[omnibus] set rpath for LC_LOAD_DYLIB and LC_ID_DYLIB when they use the build directory

### PR DESCRIPTION
**Pull request by Pythyu**
### What does this PR do?

Allow `dmg/pkg` packages to still install to `/opt/datadog-agent` even when building in a different repository.
Set the path of dylib `LC_LOAD_DYLIB` and `LC_ID_DYLIB` to use the `@rpath` instead of using the true path of the build directory.

### Motivation

Needed for macOS runners

### CODE REVIEW

File: release.json [hunk 1]
```diff
@@ -8,10 +8,10 @@
     "nightly": {
         "INTEGRATIONS_CORE_VERSION": "master",
         "OMNIBUS_SOFTWARE_VERSION": "375618d70253293d71b13f9385260aa3dedd7125",
-        "OMNIBUS_RUBY_VERSION": "db96ee7e2a9269ab864f618d7c447bc22e107fff",
+        "OMNIBUS_RUBY_VERSION": "bc5e0fb395c4b106967f97f8738d23705e5bc933",
         "JMXFETCH_VERSION": "0.49.3",
         "JMXFETCH_HASH": "258085a94d529a6bdf914db36dd50faf6fde2cebc44b1f54a60eb209a5d8917c",
-        "MACOS_BUILD_VERSION": "master",
+        "MACOS_BUILD_VERSION": "pythyu/custom_path_test",
```
Comment (Pythyu) [inline release.json:14]: WIll be reverted to `master` only used for testing

File: release.json [hunk 2]
```diff
@@ -27,10 +27,10 @@
     "nightly-a7": {
         "INTEGRATIONS_CORE_VERSION": "master",
         "OMNIBUS_SOFTWARE_VERSION": "375618d70253293d71b13f9385260aa3dedd7125",
-        "OMNIBUS_RUBY_VERSION": "db96ee7e2a9269ab864f618d7c447bc22e107fff",
+        "OMNIBUS_RUBY_VERSION": "bc5e0fb395c4b106967f97f8738d23705e5bc933",
         "JMXFETCH_VERSION": "0.49.3",
         "JMXFETCH_HASH": "258085a94d529a6bdf914db36dd50faf6fde2cebc44b1f54a60eb209a5d8917c",
-        "MACOS_BUILD_VERSION": "master",
+        "MACOS_BUILD_VERSION": "pythyu/custom_path_test",
```
Comment (Pythyu) [inline release.json:33]: WIll be reverted to `master` only used for testing

File: tasks/omnibus.py [hunk 3]
```diff
@@ -404,7 +404,23 @@ def rpath_edit(ctx, install_path, target_rpath_dd_folder, platform="linux"):
...
+            if install_path in dylib_paths:
+                for otool_line in dylib_paths.splitlines():
+                    if "name" not in otool_line:
+                        continue
+                    dylib_path = otool_line.strip().split(" ")[1]
+                    if install_path not in dylib_path:
+                        continue
+                    new_dylib_path = dylib_path.replace(install_path, "@rpath")
+                    # Edit LC_LOAD_DYLIB paths relative to rpath
+                    ctx.run(f"install_name_tool -change {dylib_path} {new_dylib_path} {file}")
```
Comment (alopezz) [inline tasks/omnibus.py:423]: `install_name_tool` seems to accept [multiple](https://keith.github.io/xcode-man-pages/install_name_tool.1.html#change) `-change` arguments on the same invokation. Would it be worth trying to group all of these changes?

**Review by alopezz — COMMENTED** [review]
The logic is starting to grow in complexity and we're approaching a point where adding tests might be a good idea.

File: tasks/omnibus.py (hunk 3 above)
Comment (Pythyu) [reply tasks/omnibus.py:423]: 😮 didn't notice it was possible will try, if yes then definitely worth 🚀

Comment (Pythyu) [reply tasks/omnibus.py:423]: as long we don't reach the shell argument limit

File: tasks/omnibus.py [hunk 4]
```diff
@@ -404,7 +404,23 @@ def rpath_edit(ctx, install_path, target_rpath_dd_folder, platform="linux"):
         else:
             if file_type != "application/x-mach-binary":
                 continue
-            binary_rpath = ctx.run(f'otool -l {file} | grep -A 2 "RPATH"', warn=True, hide=True).stdout
+            errors = ctx.run(f'otool -l {file} > tmpotoolfile', warn=True, hide=True).stderr
```
[the remark this answers is not in the document]
Comment (Pythyu) [reply tasks/omnibus.py:407]: Tbh on that point it's more me not trusting python speed to search in text especially when it's called a lot. Possible yes, but I'd prefer to rely on something fast like grep, macOS is slow enough I don't want to make it worse. 
I can quickly try some regex but I only see it as an extra build time

[the remark this answers is not in the document]
Comment (alopezz) [reply tasks/omnibus.py:407]: What I had in mind was a bit besides the use of `grep` vs Python's own regexp's (though that might still be worth testing before discounting one as faster than the other).

I was referring specifically to avoiding the extra file. Something like this:
```python
            result = ctx.run(f'otool -l {file}', warn=True, hide=True)
            if result.exited != 0:
                continue

            binary_rpath = ctx.run(
                'grep -A 2 "RPATH"', warn=True, hide=True, in_stream=io.StringIO(result.stdout)
            ).stdout
            dylib_paths = ctx.run(
                'grep -A 2 "LC_LOAD_DYLIB\|LC_ID_DYLIB"', warn=True, hide=True, in_stream=io.StringIO(result.stdout)
            ).stdout
```

File: tasks/omnibus.py (hunk 3 above)
Comment (alopezz) [reply tasks/omnibus.py:423]: > as long we don't reach the shell argument limit

I would be surprised if we did, that would mean a single binary depending on hundreds of libraries.

Comment (Pythyu) [reply tasks/omnibus.py]: Last commit reduce the time ~30 sec n oneed to compact further lets play safe and keep it as is

File: tasks/omnibus.py [hunk 5]
```diff
@@ -387,10 +388,51 @@ def manifest(
...
+
+def _replace_dylib_id_path_with_rpath(ctx, dylib_paths, install_path, file):
+    for otool_line in dylib_paths.splitlines():
+        if "name" not in otool_line:
+            continue
+        dylib_path = otool_line.strip().split(" ")[1]
+        if install_path not in dylib_path:
+            continue
+        new_dylib_path = dylib_path.replace(install_path, "@rpath")
+        ctx.run(f"install_name_tool -id {new_dylib_path} {file}")
```
Comment (alopezz) [inline tasks/omnibus.py:410]: Some ideas here, let me know what you think:

```suggestion
def _otool_install_path_replacements(otool_output, install_path):
    """Returns a mapping of path replacements from `otool -l` output
    where references to `install_path` are replaced by `@rpath`."""
    for otool_line in dylib_paths.splitlines():
        if "name" not in otool_line:
            continue
        dylib_path = otool_line.strip().split(" ")[1]
        if install_path not in dylib_path:
            continue
        new_dylib_path = dylib_path.replace(install_path, "@rpath")
        yield dylib_path, new_dylib_path


def _replace_dylib_paths_with_rpath(ctx, otool_output, install_path, file):
    for dylib_path, new_dylib_path in _otool_install_path_replacements(otool_output, install_path):
        ctx.run(f"install_name_tool -change {dylib_path} {new_dylib_path} {file}")


def _replace_dylib_id_path_with_rpath(ctx, otool_output, install_path, file):
    for dylib_path, new_dylib_path in _otool_install_path_replacements(otool_output, install_path):
        ctx.run(f"install_name_tool -id {new_dylib_path} {file}")
```

There might be other ways to slice it, the main point being to deduplicate common logic and making it more explicit that we're processing _otool_ output (`dylib_paths` can be somewhat misleading there).

**Review by alopezz — APPROVED** [review]
I have a suggestion to try to improve the structure of the code a bit, as well as a reminder that I think tests could be somewhat helpful to cover some of the logic here.

Other than that this looks fine.

Also, an additional reminder to  revert the release.json changes before merging in any case.

**Pull request merged**